### PR TITLE
ultralytics/yolov5 integration

### PR DESCRIPTION
**Pull request by bfineran**

### CODE REVIEW

**Comment by bfineran** [comment]
this first commit includes requirements for running recipes within the yolov5 training script as well as a SparseZoo integration for recipes and checkpoint files.  testing is WIP and is blocked by other resource priorities

File: examples/ultralytics-sparseml/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+cd yolov5
+
+# install dependencies
+pip install -r requirements.txt
+pip install sparseml
+```
+
+
+## Script
+`examples/timm-sparseml/main.py` modifies
```
Comment (mgoin) [inline examples/ultralytics-sparseml/README.md:51]: should this be `examples/ultralytics-sparseml/main.py`?

File: examples/ultralytics-sparseml/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+[`train.py`](https://github.com/ultralytics/yolov5/blob/master/train.py)
+from yolov5 to include a `sparseml-recipe-path` argument
+to run SparseML optimizations with.  This can be a file path to a local
+SparseML recipe or a SparseZoo model stub prefixed by `zoo:` such as
+`zoo:cv/detection/yolo_v3-spp/pytorch/ultralytics/coco/pruned-aggressive`.
```
Comment (mgoin) [inline examples/ultralytics-sparseml/README.md:56]: i didn't know you could use the model stub as recipe path, that's really nice

**Review by mgoin — DISMISSED** [review]
cool, docs and integration looks really clean to me. accepting now assuming that we'll have testing added in later work

File: examples/ultralytics-sparseml/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# SparseML-ultralytics/yolov5 integration
+This directory provides a SparseML integrated training script for the popular
+[ultralytics/yolov5](https://github.com/ultralytics/yolov5)
+repository also known as [timm](https://pypi.org/project/timm/).
```
Comment (markurtz) [inline examples/ultralytics-sparseml/README.md:20]: left over timm reference

File: examples/ultralytics-sparseml/main.py [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+    logger.info('Optimizer groups: %g .bias, %g conv.weight, %g other' % (len(pg2), len(pg1), len(pg0)))
+    del pg0, pg1, pg2
+
+    # Scheduler https://arxiv.org/pdf/1812.01187.pdf
+    # https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#OneCycleLR
+    if opt.linear_lr:
+        lf = lambda x: (1 - x / (epochs - 1)) * (1.0 - hyp['lrf']) + hyp['lrf']  # linear
+    else:
+        lf = one_cycle(1, hyp['lrf'], epochs)  # cosine 1->hyp['lrf']
+    scheduler = lr_scheduler.LambdaLR(optimizer, lr_lambda=lf)
```
Comment (markurtz) [inline examples/ultralytics-sparseml/main.py:146]: If the recipe is overriding the LR, we should likely kill the scheduler here

**Review by markurtz — CHANGES_REQUESTED** [review]
Same overall comments as the rwightman integration:
- migrate it over to your suggestion of having a folder for integrations
- name the root just ultralytics since we're already in the sparseml repo
- remove path from the arg since this supports both paths and stubs for sparseml-recipe-path
- change out functionality for sparse-transfer-learn to be more generic for sparsezoo-recipe-type
- issues with using min_epoch from manager
- logging for when we're overriding something such as with the number of epochs

**Comment by markurtz** [comment]
Oh also, can we get the exact checkpoint of the integrated repo this was tested with in the even people have issues with newer versions?

**Comment by bfineran** [comment]
Thanks for review @markurtz @mgoin should be up to date with reviews now

**Review by mgoin — DISMISSED** [review]
LGTM

File: integrations/ultralytics/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+cd yolov5
+
+# install dependencies
+pip install -r requirements.txt
+pip install sparseml
+```
+
+
+## Script
+`integrations/ultralytics/main.py` modifies
```
Comment (markurtz) [inline integrations/ultralytics/README.md:51]: One more small change that I missed, can we rename this file to train.py? This way there is a one to one mapping without a lookup of names for the file(s) in the integration repo to the files stored in sparseml

**Review by markurtz — CHANGES_REQUESTED** [review]

**Review by markurtz — DISMISSED** [review]

**Review by mgoin — DISMISSED** [review]

File: integrations/ultralytics/train.py [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+    ####################################################################################
+    parser.add_argument(
+        "--sparseml-recipe",
+        required=True,
+        type=str,
+        help="path to a SparseML recipe file or a SparseZoo model stub for a recipe to load. "
+             "SparseZoo stubs should be preceded by 'zoo:'. i.e. '/path/to/local/recipe.yaml', "
+             "'zoo:zoo/model/stub'"
+    )
+    parser.add_argument('--weights', type=str, default='yolov5s.pt', help='initial weights path')
```
Comment (bfineran) [inline integrations/ultralytics/train.py:510]:
```suggestion
```

File: integrations/ultralytics/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+This directory provides a SparseML integrated training script for the popular
+[ultralytics/yolov5](https://github.com/ultralytics/yolov5)
+repository.
+
+Using this integration, you will be able to apply SparseML optimizations
+to the powerful training flows provided in the yolov5 repository.
+
+Some of the tasks you can perform using this integration include, but are not limited to:
+* model pruning
+* quantization-aware-training
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:27]:
```suggestion
* quantization-aware training
```

File: integrations/ultralytics/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+* sparse quantization-aware-training
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:28]:
```suggestion
* sparse quantization-aware training
```

File: integrations/ultralytics/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+* sparse transfer learning
+
+## Installation
+To use both the script, clone both repositories, install their dependencies,
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:32]:
```suggestion
To use the scripts, clone both repositories, install their dependencies,
```

File: integrations/ultralytics/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+follow the normal yolov5 training flow with the given SparseML optimizations enabled.
+
+Some considerations:
+
+* `--sparseml-recipe` is a required parameter
+* `--epochs` will now be overridden by the epochs set in the SparseML recipe
+* if using learning rate schedulers both with the yolov5 script and your recipe, they
+may conflict with each other causing unintended side effects, choose
+hyperparameters accordingly.
+* Modifiers will log their outputs to the console as well as to the tensorboard file
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:73]:
```suggestion
* Modifiers will log their outputs to the console as well as to the TensorBoard file
```

File: integrations/ultralytics/README.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+Running the script will
+follow the normal yolov5 training flow with the given SparseML optimizations enabled.
+
+Some considerations:
+
+* `--sparseml-recipe` is a required parameter
+* `--epochs` will now be overridden by the epochs set in the SparseML recipe
+* if using learning rate schedulers both with the yolov5 script and your recipe, they
+may conflict with each other causing unintended side effects, choose
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:71]:
```suggestion
may conflict with each other causing unintended side effects, so choose
```

File: integrations/ultralytics/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+hyperparameters accordingly.
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:72]:
```suggestion
hyperparameters accordingly
```

**Review by markurtz — DISMISSED** [review]

File: integrations/ultralytics/README.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+The latest commit hash that `train.py` is based on is included in the docstring.
+
+
+#### Example Command
+Call the script from the `yolov5` directory, passing in the same arguments as
+`train.py`, with the additional SparseML argument(s) included.
+```bash
+python train.py \
+  --sparseml-recipe /PATH/TO/RECIPE/recipe.yaml \
+  <regular yolov5/train.py paramters>
```
Comment (jeanniefinks) [inline integrations/ultralytics/README.md:94]:
```suggestion
  <regular yolov5/train.py parameters>
```

**Review by jeanniefinks — DISMISSED** [review]
minor typo fix from me per last comment and you are gtg from me.

**Review by mgoin — DISMISSED** [review]

File: integrations/ultralytics/train.py [hunk 14]
```diff
@@ -0,0 +1,728 @@
...
+    try:
+        import wandb
+    except ImportError:
+        wandb = None
+        prefix = colorstr('wandb: ')
+        logger.info(f"{prefix}Install Weights & Biases for YOLOv5 logging with 'pip install wandb' (recommended)")
+    if not opt.evolve:
+        tb_writer = None  # init loggers
+        if opt.global_rank in [-1, 0]:
+            logger.info(f'Start Tensorboard with "tensorboard --logdir {opt.project}", view at http://localhost:6006/')
```
Comment (jeanniefinks) [inline integrations/ultralytics/train.py:644]:
```suggestion
            logger.info(f'Start TensorBoard with "tensorboard --logdir {opt.project}", view at http://localhost:6006/')
```

**Review by markurtz — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
